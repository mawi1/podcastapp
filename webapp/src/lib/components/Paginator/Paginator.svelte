<script module lang="ts">
  export function getPaginationInfo(
    searchParams: URLSearchParams,
    totalItems: number,
    itemsPerPage: number
  ):
    | {
        previous: string | undefined;
        next: string | undefined;
        links: {
          page: number;
          href: string;
          isActive: boolean;
        }[];
      }
    | undefined {
    if (totalItems <= itemsPerPage) {
      return undefined;
    }

    let currentPage = Number(searchParams.get("page")) || 1;
    const totalPages = Math.ceil(totalItems / itemsPerPage);
    if (currentPage > totalPages) {
      return undefined;
    }

    const PAGINATOR_LENGTH = 5;
    const PAGES_LEFT_RIGHT = 2;
    let start, end;
    if (currentPage - PAGES_LEFT_RIGHT < 1) {
      start = 1;
      end = Math.min(PAGINATOR_LENGTH, totalPages);
    } else if (currentPage + PAGES_LEFT_RIGHT > totalPages) {
      start = Math.max(1, totalPages - PAGINATOR_LENGTH + 1);
      end = totalPages;
    } else {
      start = currentPage - PAGES_LEFT_RIGHT;
      end = currentPage + PAGES_LEFT_RIGHT;
    }

    const makeParams = (n: number) => {
      const newParams = new URLSearchParams(searchParams);
      newParams.set("page", n.toString());
      return `?${newParams.toString()}`;
    };

    let links = [];
    for (let p = start; p <= end; p++) {
      links.push({ href: makeParams(p), page: p, isActive: p === currentPage });
    }

    return {
      previous: currentPage === 1 ? undefined : makeParams(currentPage - 1),
      next: currentPage === totalPages ? undefined : makeParams(currentPage + 1),
      links,
    };
  }
</script>

<script lang="ts">
  import { page } from "$app/state";

  import { PAGE_SIZE } from ".";

  const { totalItems }: { totalItems: number } = $props();

  let paginationInfo = $derived(getPaginationInfo(page.url.searchParams, totalItems, PAGE_SIZE));
</script>

{#if paginationInfo}
  <div class="@container">
    <nav class="mt-6 flex justify-between @min-[550px]:justify-center">
      {#if paginationInfo.previous}
        <a class="btn" href={paginationInfo.previous} aria-label="previous"
          ><i class="fa-solid fa-caret-left"></i></a
        >
      {:else}
        <div class="w-12"></div>
      {/if}

      <div class="hidden @min-[340px]:block @min-[550px]:mx-6">
        {#each paginationInfo.links as { page, href, isActive }}
          <a
            class:font-extrabold={isActive}
            class="inline-block px-2 @min-[460px]:px-2.5 py-1.5"
            {href}>{page}</a
          >
        {/each}
      </div>

      {#if paginationInfo.next}
        <a class="btn" href={paginationInfo.next} aria-label="next"
          ><i class="fa-solid fa-caret-right"></i></a
        >
      {:else}
        <div class="w-12"></div>
      {/if}
    </nav>
  </div>
{/if}

<style>
  @reference "tailwindcss";

  .btn {
    @apply flex w-12 items-center justify-center rounded border border-zinc-400 py-2;
  }
</style>
