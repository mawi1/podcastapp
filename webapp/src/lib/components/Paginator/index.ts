export const PAGE_SIZE = 50;

export { default } from "./Paginator.svelte";
