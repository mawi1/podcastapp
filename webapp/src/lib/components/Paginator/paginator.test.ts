import { describe, expect, test } from "vitest";

import { getPaginationInfo } from "./Paginator.svelte";

describe("getPaginationInfo", () => {
  test("is undefined if all items fit on one page", () => {
    const p = getPaginationInfo(new URLSearchParams(), 10, 10);
    expect(p).toBeUndefined();
  });

  test("is defined if all items do not fit on one page", () => {
    const p = getPaginationInfo(new URLSearchParams(), 11, 10);
    expect(p).toBeDefined();
  });

  test("is undefined page > totalpages", () => {
    const p = getPaginationInfo(new URLSearchParams("page=19"), 337, 19);
    expect(p).toBeUndefined();
  });

  test("previous is undefined on page 1", () => {
    const p = getPaginationInfo(new URLSearchParams("page=1"), 587, 10);
    expect(p!.previous).toBeUndefined();
  });

  test("next is undefined on last page", () => {
    const p = getPaginationInfo(new URLSearchParams("page=82"), 983, 12);
    expect(p!.next).toBeUndefined();
  });

  test.each([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11])("current page (%i) is active is true", (page) => {
    const p = getPaginationInfo(new URLSearchParams(`page=${page}`), 101, 10);
    expect(p!.links.find((e) => e.isActive)!.page).toBe(page);
  });

  test.each([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11])("only one page (%i) is active is true", (page) => {
    const p = getPaginationInfo(new URLSearchParams(`page=${page}`), 101, 10);
    expect(p!.links.filter((e) => e.isActive).length).toBe(1);
  });

  test("no param -> page 1", () => {
    const p = getPaginationInfo(new URLSearchParams(), 613, 25);
    expect(p!.links[0].page).toBe(1);
  });

  test.each([
    ["page=1", 867, 16, [1, 2, 3, 4, 5]],
    ["page=2", 867, 16, [1, 2, 3, 4, 5]],
    ["page=32", 867, 16, [30, 31, 32, 33, 34]],
    ["page=41", 867, 16, [39, 40, 41, 42, 43]],
    ["page=54", 867, 16, [51, 52, 53, 54, 55]],
    ["page=55", 867, 16, [51, 52, 53, 54, 55]],
  ])("page numbers are correct for %s", (searchParams, totalItems, itemsPerPage, expected) => {
    const p = getPaginationInfo(new URLSearchParams(searchParams), totalItems, itemsPerPage);
    expect(p!.links.map((l) => l.page)).toStrictEqual(expected);
  });

  test.each([
    ["page=1", 1495, 44, "?page=2"],
    ["page=19", 1495, 44, "?page=20"],
    ["page=28", 1495, 44, "?page=29"],
  ])("next is correct for %s", (searchParams, totalItems, itemsPerPage, expected) => {
    const p = getPaginationInfo(new URLSearchParams(searchParams), totalItems, itemsPerPage);
    expect(p!.next).toBe(expected);
  });

  test.each([
    ["page=12", 5378, 57, "?page=11"],
    ["page=74", 5378, 57, "?page=73"],
    ["page=95", 5378, 57, "?page=94"],
  ])("prev is correct for %s", (searchParams, totalItems, itemsPerPage, expected) => {
    const p = getPaginationInfo(new URLSearchParams(searchParams), totalItems, itemsPerPage);
    expect(p!.previous).toBe(expected);
  });

  test.each([
    ["page=6", 3270, 25, ["?page=4", "?page=5", "?page=6", "?page=7", "?page=8"]],
    ["page=101", 3270, 25, ["?page=99", "?page=100", "?page=101", "?page=102", "?page=103"]],
    ["page=69", 3270, 25, ["?page=67", "?page=68", "?page=69", "?page=70", "?page=71"]],
  ])("hrefs are correct for %s", (searchParams, totalItems, itemsPerPage, expected) => {
    const p = getPaginationInfo(new URLSearchParams(searchParams), totalItems, itemsPerPage);
    expect(p!.links.map((l) => l.href)).toStrictEqual(expected);
  });

  test("other search params are keept", () => {
    const p = getPaginationInfo(new URLSearchParams("page=3&foo=bar"), 581, 23);
    expect(p).toMatchSnapshot();
  });
});
