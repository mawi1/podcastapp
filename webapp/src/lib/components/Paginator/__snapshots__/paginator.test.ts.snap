// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getPaginationInfo > other search params are keept 1`] = `
{
  "links": [
    {
      "href": "?page=1&foo=bar",
      "isActive": false,
      "page": 1,
    },
    {
      "href": "?page=2&foo=bar",
      "isActive": false,
      "page": 2,
    },
    {
      "href": "?page=3&foo=bar",
      "isActive": true,
      "page": 3,
    },
    {
      "href": "?page=4&foo=bar",
      "isActive": false,
      "page": 4,
    },
    {
      "href": "?page=5&foo=bar",
      "isActive": false,
      "page": 5,
    },
  ],
  "next": "?page=4&foo=bar",
  "previous": "?page=2&foo=bar",
}
`;
